github/ratelimit_test.go

```diff
@@ -0,0 +1,331 @@
+		headers:    abuseHeaders(defaultAbuseReset),
+		response:   apiResponse{Data: "whatever"},
+		err:        nil,
+	},
+	// 403 with proper AbuseRateLimit body but without headers
+	"/abuse_no_headder": response{
+		statusCode: http.StatusForbidden,
+		headers:    nil,
+		response: apiResponse{
+			apiErrorResponse: apiErrorResponse{Message: "Found an abuse detection mechanism"},
+		},
+		err: nil,
+	},
+	// 403 without proper AbuseRateLimit body nor headers
+	"/forbidden_403": response{
+		statusCode: http.StatusForbidden,
+		headers:    nil,
+		response:   apiResponse{Data: "forbidden"},
+		err:        nil,
+	},
+	// returns a network error
+	"/network_error": response{
+		statusCode: http.StatusTeapot,
+		headers:    nil,
+		response:   apiResponse{},
+		err:        fmt.Errorf("network error"),
+	},
+}
+
+type response struct {
+	statusCode int
+	headers    func(time.Time) map[string]string
+	response   apiResponse
+	err        error
+}
+
+type apiResponse struct {
+	apiErrorResponse
+	Data string `json:"data,omitempty"`
+}
+
+func rateLimitHeaders(wait time.Duration) func(time.Time) map[string]string {
+	return func(when time.Time) map[string]string {
+		return map[string]string{
+			"X-RateLimit-Reset":     strconv.FormatInt(when.Add(wait).Unix(), 10),
+			"X-RateLimit-Remaining": "0",
+		}
+	}
+}
+
+func abuseHeaders(wait time.Duration) func(time.Time) map[string]string {
+	return func(when time.Time) map[string]string {
+		return map[string]string{
+			"Retry-After": strconv.FormatInt(int64(wait.Seconds()), 10),
+		}
+	}
+}
+
+func newResponse(content apiResponse, headers func(time.Time) map[string]string, statusCode int) (*http.Response, error) {
+	data, err := json.Marshal(content)
+	if err != nil {
+		return nil, err
+	}
+
+	respWriter := httptest.NewRecorder()
+	respWriter.WriteHeader(statusCode)
+	_, err = respWriter.Write(data)
+	if err != nil {
+		return nil, err
+	}
+
+	response := respWriter.Result()
+
+	if headers != nil {
+		for k, v := range headers(time.Now()) {
+			response.Header.Set(k, v)
+		}
+	}
+
+	return response, err
+}
+
+func newRequest(url string) *http.Request {
+	req, _ := http.NewRequest("GET", url, http.NoBody)
+	return req
+}
+
+// gitHubTransportResponseMock avoids calling the network to get the response directly from testcases
+type gitHubTransportResponseMock struct{}
+
+func (gh *gitHubTransportResponseMock) RoundTrip(req *http.Request) (*http.Response, error) {
+	url := req.URL.String()
+	tc, ok := testCases[url]
+	if !ok {
+		return newResponse(apiResponse{}, nil, http.StatusOK)
+	}
+
+	if tc.err != nil {
+		return &http.Response{}, tc.err
+	}
+
+	return newResponse(tc.response, tc.headers, tc.statusCode)
+}
+
+func TestRateLimitSuite(t *testing.T) {
+	suite.Run(t, new(RateLimitSuite))
+}
+
+type RateLimitSuite struct {
+	suite.Suite
+	transport  *RateLimitTransport
+	loggerMock *testutils.LoggerMock
+}
+
+func (s *RateLimitSuite) SetupSuite() {
+	defaultAbuseRetryAfter = default403Reset
```

Comment:
when you a change global variable in `Setup` you need to return it to the previous value in `Down`. Otherwise later, somebody (or maybe yourself) will have a very hard time debugging why their new test doesn't work as expected and returns different results depends on if all the test suite was executed or only the new test.

Reply:
that's a pretty good point, with 55bdde4 globals are no longer used to store state